### PR TITLE
perf: Reduce copies in ToCompactStr for integer types

### PR DESCRIPTION
**Pull request by ParkMyCar**
This PR improves the performance of the `trait ToCompactStr` by reducing the number of copies necessary when converting an integer.

Locally I see at least 15% - 46% improvements

### CODE REVIEW

File: compact_str/src/traits.rs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+/// if the `Display` implementation returns an error.
+/// This indicates an incorrect `Display` implementation
+/// since `std::fmt::Write for String` never returns an error itself and
+/// the implementation of `ToCompactStr::to_compact_str` only panic if the `Display`
+/// implementation is incorrect..
+impl<T: fmt::Display> ToCompactStr for T {
+    #[inline]
+    fn to_compact_str(&self) -> CompactStr {
+        let repr = match_type!(self, {
+            u8 as s => s.into_repr(),
```
Comment (NobodyXu) [inline compact_str/src/traits.rs:64]: `&self` cannot be `u8`, it can only be a reference.

File: compact_str/src/traits.rs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            &u8 as s => s.into_repr(),
+            i8 as s => s.into_repr(),
```
Comment (NobodyXu) [inline compact_str/src/traits.rs:66]: Same for all the other types here, it cannot be take by value, but by reference only.

**Review by NobodyXu — CHANGES_REQUESTED** [review]

**Comment by NobodyXu** [comment]
@ParkMyCar According to my understanding this PR optimize `ToCompactStr` specialisation for integers, by directly format `u8`, `i8`, `u16`, `i16`, `u32` and `i32` into the `CompactStr` inline storage directly.

For `u64` and `i64`, they are formatted into `CompactStr` inline storage on 64-bit platform.

For `u128` and `i128`, they are formatted using `std::format!` and might require heap allocation.
But doesn't that mean formatting of 128-bit integer could be much slower than other integer types?

File: compact_str/src/traits.rs (hunk 1 above)
Comment (NobodyXu) [reply compact_str/src/traits.rs:64]: Note that it could be `&&u8` or `&&&u8`, but not sure it is worth the effort to optimize for these cases.

Comment (ParkMyCar) [reply compact_str/src/traits.rs]: Ah that makes sense, thanks for the help here!

Comment (ParkMyCar) [reply compact_str/src/traits.rs]: Ah that makes sense, thanks for the help here!

**Comment by ParkMyCar** [comment]
> @ParkMyCar According to my understanding this PR optimize `ToCompactStr` specialisation for integers, by directly format `u8`, `i8`, `u16`, `i16`, `u32` and `i32` into the `CompactStr` inline storage directly.
> 
> For `u64` and `i64`, they are formatted into `CompactStr` inline storage on 64-bit platform.
> 
> For `u128` and `i128`, they are formatted using `std::format!` and might require heap allocation. But doesn't that mean formatting of 128-bit integer could be much slower than other integer types?

I updated the implementation, 8, 16, and 32 bit integers should always be inlined, 64 bit integers will always be inlined on 64-bit machines, and possibly on 32-bit machines, depending on how many digits the number has.

Good call with the 128-bit integers. I re-added the dependency on `itoa` and use that for the impl of `ToCompactStr`. It turns out determining the str length of a 128 bit int takes longer than the memcpy, so we use your original impl of writing to a `itoa::Buffer` and then copying

**Comment by ParkMyCar** [comment]
> @ParkMyCar According to my understanding this PR optimize `ToCompactStr` specialisation for integers, by directly format `u8`, `i8`, `u16`, `i16`, `u32` and `i32` into the `CompactStr` inline storage directly.
> 
> For `u64` and `i64`, they are formatted into `CompactStr` inline storage on 64-bit platform.
> 
> For `u128` and `i128`, they are formatted using `std::format!` and might require heap allocation. But doesn't that mean formatting of 128-bit integer could be much slower than other integer types?

I updated the implementation, 8, 16, and 32 bit integers should always be inlined, 64 bit integers will always be inlined on 64-bit machines, and possibly on 32-bit machines, depending on how many digits the number has.

Good call with the 128-bit integers. I re-added the dependency on `itoa` and use that for the impl of `ToCompactStr`. It turns out determining the str length of a 128 bit int takes longer than the memcpy, so we use your original impl of writing to a `itoa::Buffer` and then copying

**Pull request merged**